atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ReloadingCloseableContainer.java

```diff
@@ -0,0 +1,222 @@
+     * After this method has returned, no further resources will be created in this container, and all subsequent
+     * {@link #runWithResource} invocations will fail with a {@link SafeIllegalStateException}.
+     *
+     * Note: This method does <b>block</b> on any inflight {@link #runWithResource} invocations.
+     *
+     * Any {@link IOException} from closing the underlying resource will not be propagated.
+     *
+     * See {@link Closeable#close()}.
+     */
+    @Override
+    public void close() {
+        Lock isClosedWriteLock = isClosedLock.writeLock();
+        runWithLock(isClosedWriteLock, () -> {
+            isClosed = true;
+            currentResourceWithLock.get().ifPresent(this::closeResourceAndLock);
+            refreshableSubscriptionDisposable.dispose();
+        });
+    }
+
+    /***
+     * Returns the result of `task` applied with the current resource.
+     * This resource is guaranteed not to be closed within scope of this function <i>by this container</i>.
+     * There are no guarantees on the resource being closed externally e.g. if there is a shared reference, nor if the
+     * resource is closed within scope by the task itself.
+     *
+     * If an update to the refreshable occurs during an invocation, any <i>inflight</i> invocations will <i>not</i> use
+     * the new resource.
+     *
+     * <b>Do not</b> block on close or update the refreshable within the scope of runWith. Doing so will result in a
+     * deadlock.
+     *
+     * There are no timeouts applied to these invocations. If required, consider using the
+     * {@link com.google.common.util.concurrent.TimeLimiter} utilities. Long-running invocations will block close.
+     *
+     * @param task The function to execute using the current resource. This function may throw a checked exception,
+     *             and this exception will propagate.
+     * @param <V> The return type of the task
+     * @param <K> The checked exception type, if any. If multiple checked exceptions are thrown, this will resolve to
+     *          the lowest common ancestor, which may be {@link Exception}
+     * @return The result of `task`
+     * @throws K Any checked exceptions thrown by the underlying task.
+     */
+    public <V, K extends Exception> V runWithResource(FunctionCheckedException<T, V, K> task) throws K {
+        return runIfNotClosed(() -> {
+            Pair<ReadWriteLock, T> resourceWithLock = refreshableResourceWithLock.get();
+            Lock readLock = resourceWithLock.getLhSide().readLock();
+            return runWithLock(readLock, () -> task.apply(resourceWithLock.getRhSide()));
+        });
+    }
+
+    /**
+     * Utility method for {@link #runWithResource} for tasks that have no return value.
+     *
+     * See {@link #runWithResource} for semantics, guarantees and conditions.
+     *
+     */
+    public <K extends Exception> void executeWithResource(ConsumerCheckedException<T, K> task) throws K {
```

Comment:
I'm not particularly tied to the names, I chose runWithResource because `runWith` does make the usage obvious and is consistent with other parts of the codebase e.g `LockWatchEventLog` (although could be `runXWith` to match the API), and `executeWith` because that's similar to the `Executor` API for callables that don't return any thing.

I didn't overload `runWith` because I was already getting ambiguous usage when writing tests, and can imagine that being annoying in the wild. This separation, to me, was clearer, but happy to discuss this.